src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -1229,23 +1233,126 @@ private int SkipBlockComment(int i)
+                case 'r': return "\r";
+                case 't': return "\t";
+                case 'u': return ScanUnicodeEscapeSequence();
+                case 'v': return "\v";
+                default: return c.ToString();
+            }
+        }
+
+        private string ScanUnicodeEscapeSequence()
+        {
+            int escSeqStartIndex = _currentIndex - 2;
+            int maxNumberOfHexDigits = 6;
+            bool isTerminated = false;
+
+            char c = GetChar();
+            if (c != '{')
+            {
+                UngetChar();
+
+                IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                return s_unknownUnicodeChar;
+            }
+
+            // Scan hex chars after the Unicode escape sequence start.
+            var sb = GetStringBuilder();
+            int i;
+            for (i = 0; i < maxNumberOfHexDigits; i++)
+            {
+                c = GetChar();
+
+                // Sequence has been terminated.
+                if (c == '}')
+                {
+                    if (i == 0)
+                    {
+                        // Sequence must have at least one hex char.
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                        return s_unknownUnicodeChar;
+                    }
+
+                    isTerminated = true;
+                    break;
+                }
+
+                if (!c.IsHexDigit())
+                {
+                    UngetChar();
+
+                    ReportError(_currentIndex, () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+
+                sb.Append(c);
+            }
+
+            if (!isTerminated)
+            {
+                c = GetChar();
+                if (c != '}')
+                {
+                    UngetChar();
+                    ReportError(_currentIndex,
+                        i == maxNumberOfHexDigits && c.IsHexDigit()
+                            ? (Expression<Func<string>>)(() => ParserStrings.TooManyDigitsInUnicodeEscapeSequence)
+                            : () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+            }
+
+            string hexStr = GetStringAndRelease(sb);
+
+            int unicodeValue = int.Parse(hexStr, NumberStyles.AllowHexSpecifier, NumberFormatInfo.InvariantInfo);
+            if (unicodeValue <= Char.MaxValue)
+            {
+                return ((char)unicodeValue).ToString();
+            }
+            else if (unicodeValue <= 0x10FFFF)
+            {
+                return Char.ConvertFromUtf32(unicodeValue);
```

Comment:
As for the **current behavior in other shells:**

Yes, I didn't mean to imply that (a) they all produce an _error_ and (b) fail at _parsing_ - with the exception of Ruby v2.4+, which seems to have made an explicit decision to disallow surrogate code points (unlike earlier versions). You'll see that the `python` command (2.x) does error on _output_ if you send the output somewhere other _than the terminal_ (which is why my code had a trailing `| cat`), but the main reason is that it's trying to convert to _ASCII_.  

Python 3.x provides a more relevant pointer (it breaks even when you print to the terminal):  
`UnicodeEncodeError: 'utf-8' codec can't encode characters in position 0-1: surrogates not allowed`.

Pragmatically speaking, even if you get to send them to a file, for instance, _they're useless in UTF-8_ - see next point.

---

As for **whether it makes sense to support surrogate pairs:**

_Surrogate pairs are purely a UTF-16 concept, and they do not apply to UTF-8, which will be the typical _output_ encoding_.

Without knowing all the intricacies:

* Presumably, that's why languages such as Python 3.x and Perl (since at least v5.18) break on _output_, when attempting to _convert_ surrogate pairs to _UTF-8_ - because  _surrogate code points have no UTF-8 representation_.

* That's why Ruby 2.4+, more proactively - and more sensibly, in my view - breaks at _parsing_ time, disallowing the use of surrogate code points in Unicode escape sequences altogether.

Maybe we won't have that problem, because it isn't _bytes_ that are being passed around, and based on a proper UTF-16 encoded string, you can, of course, convert properly to UTF-8.

How does that fit into plans for a binary pipe, though?

It seems to me that there's little to be gained, and obscure failures are lurking.

> The basic ask here is that folks want Emoji support. 

But, as you state, they do get that with `` `u{1F44d} `` - no need to resort to the more complex surrogate-pair representation.

At least it works on _Unix_ - I can't seem to be able to get it to work on _Windows_ - with or without surrogate pairs (tried with 2 fonts).

> Besides there is nothing to stop you from doing this today using sub-expressions [e.g., `$([char]0xD83d)`]

And that is what I suggest users who truly need an explicit representation of surrogate code points do.

---

**The standards / conceptual view:**

To recap my review comment: there's no such thing as as a _surrogate character_, and `U+1F44D` is simply a (symbol) character that happens to lie outside the BMP (basic multi-lingual plane), due to having a code-point value > `0xFFFF`.  
 
_Solely for the benefit of UTF-16_, this single character has an alternate representation via a _pair_ of _surrogate code points_ (which, individually, are _code points_, but not (assigned) _characters_).

Surrogate code points are of type _Surrogate_ (one of the 7 defined types; in full: _Graphic_, _Format_, _Control_, _Private-use_, _Surrogate_, _Noncharacter_, _Reserved_), and [the Unicode standard describes them in brief](http://www.unicode.org/versions/Unicode10.0.0/ch02.pdf) as follows (emphasis added):

> **Permanently reserved for UTF-16**; restricted interchange

Furthermore (emphasis added):

> **Surrogates** and Noncharacters are assigned code points, but are **not assigned to abstract characters**.

Similarly of interest is the "restricted interchange" annotation, which refers to (emphasis added):

> Code points that are not assigned to abstract characters are **subject to restrictions in interchange**:
>    Surrogate code points **cannot be conformantly interchanged using Unicode encoding forms**. They do not correspond to Unicode scalar values and thus **do not have well-formed representations in any Unicode encoding form**.

**That PowerShell / .NET use UTF-16 _internally_ to represent strings should be considered an _implementation detail_ and there's no need to reflect this detail in the Unicode escape-sequence syntax.**